### ISSUE TITLE
In-process transport deadlock during shutdown

### ISSUE THREAD

**Issue by dconnelly** [grpc/grpc-java#696]
Simultaneously shutting down both server and client sharing the same in-process transport can lead to a deadlock. During server shutdown, the transport lock is held while calling transportShutdown on the channel listener, which attempts to lock the channel. At the same time, channel.shutdownNow() holds the channel lock while also trying to lock the transport which leads to a deadlock:
```
Found one Java-level deadlock:
=============================
"AccountServer STOPPING":
  waiting to lock monitor 0x00007f88221d72a8 (object 0x000000076eb28a20, a io.grpc.ChannelImpl),
  which is held by "main"
"main":
  waiting to lock monitor 0x00007f8824015488 (object 0x000000076c2afb38, a io.grpc.transport.inprocess.InProcessTransport),
  which is held by "AccountServer STOPPING"

Java stack information for the threads listed above:
===================================================
"AccountServer STOPPING":
	at io.grpc.ChannelImpl$TransportListener.transportShutdown(ChannelImpl.java:281)
	- waiting to lock <0x000000076eb28a20> (a io.grpc.ChannelImpl)
	at io.grpc.transport.inprocess.InProcessTransport.notifyShutdown(InProcessTransport.java:151)
	- locked <0x000000076c2afb38> (a io.grpc.transport.inprocess.InProcessTransport)
	at io.grpc.transport.inprocess.InProcessTransport.shutdown(InProcessTransport.java:140)
	- locked <0x000000076c2afb38> (a io.grpc.transport.inprocess.InProcessTransport)
	at io.grpc.ServerImpl$ServerListenerImpl.serverShutdown(ServerImpl.java:240)
	- locked <0x000000076bfe81a8> (a io.grpc.ServerImpl)
	at io.grpc.transport.inprocess.InProcessServer.shutdown(InProcessServer.java:77)
	- locked <0x000000076be7fdc0> (a io.grpc.transport.inprocess.InProcessServer)
	at io.grpc.ServerImpl.shutdown(ServerImpl.java:135)
	- locked <0x000000076bfe81a8> (a io.grpc.ServerImpl)
	at com.pexlabs.grpc.AbstractGrpcServer.shutDown(AbstractGrpcServer.java:42)
	at com.google.common.util.concurrent.AbstractIdleService$2$2.run(AbstractIdleService.java:69)
	at com.google.common.util.concurrent.Callables$3.run(Callables.java:95)
	at java.lang.Thread.run(Thread.java:745)
"main":
	at io.grpc.transport.inprocess.InProcessTransport.shutdown(InProcessTransport.java:136)
	- waiting to lock <0x000000076c2afb38> (a io.grpc.transport.inprocess.InProcessTransport)
	at io.grpc.ChannelImpl.shutdown(ChannelImpl.java:128)
	- locked <0x000000076eb28a20> (a io.grpc.ChannelImpl)
	at io.grpc.ChannelImpl.shutdownNow(ChannelImpl.java:149)
	- locked <0x000000076eb28a20> (a io.grpc.ChannelImpl)
	at com.pexlabs.grpc.AbstractGrpcClient.close(AbstractGrpcClient.java:25)
	at com.pexlabs.test.TestUtil$1.doStop(TestUtil.java:25)
	at com.google.common.util.concurrent.AbstractService.stopAsync(AbstractService.java:204)
	at com.google.common.util.concurrent.ServiceManager.stopAsync(ServiceManager.java:327)
	at com.pexlabs.test.TestContext.after(TestContext.java:75)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:50)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)

Found 1 deadlock.
```

**Comment by ejona86** [comment]
Thanks for the clear bug report. That makes fixing it easier :-)

It seems the best approach is to make sure not to call `activeTransport.shutdown();` in `ChannelImpl` while the lock is held. In general, we don't hold the lock when calling the transport to avoid this sort of issue. Getting that fixed looks like it should be easy, at least at first glance.

Aside: There is a case where we do intentionally lock while calling the transport in `obtainActiveTransport()`, but we are guaranteed the transport hasn't started, so it isn't really dangerous (other than reentrancy).

**Comment by ejona86** [comment]
Fixed by #698
